state.go

```diff
@@ -21,6 +21,97 @@ import (
+//                                |    +-----------------------------+
+//   update commit                |    |                             |
+//                                v    v                             |
+//         |                                (no update)
+//         +---------------> update check ---------------->  update check wait
+//
+//                                |
+//                                | (update ready)
+//                                v
+//
+//                           update fetch
+//
+//                                |
+//                                | (update fetched)
+//                                v
+//
+//                          update install
+//
+//                                |
+//                                | (update installed,
+//                                |  enabled)
+//                                |
+//                                v
+//
+//                              reboot
+//
+//                                |
+//                                v
+//
+//                              final (daemon exit)
+//
+// Errors and their context are captured in Error states. Non-update states
+// transition to an ErrorState, while update related states (fetch, install,
+// commit) transition to UpdateErrorState that captures additional update
+// context information. Error states implement IsFatal() method to check whether
+// the cause is fatal or not.
+//
+//          +------------------> init <--------------------------------+
+//          |                                                          |
+//          |                      |                                   |
+//          |                      |                                   |
+//          |                      |                                   |
+//          |                      v                                   |
+//                                                                     |
+//     error state <--------- non-update state                         |
+//                                                                     |
+//          |                       |                                  |
+//          |                       |  (starts with                    |
+//          |                       |   update fetch)                  |
+//          |                       |                                  |
+//          |                       v                                  |
+//          |
+//          |                 update state -------------------> update error state
+//          |
+//          |                       |                                  |
+//          |                       |                                  |
+//          |                       v                                  |
+//          |                                                          |
+//          +-------------------> final (daemon exit) <----------------+
+//
```

Comment:
I think we have to make a distinction in the state diagram, "running committed update" or not. Because it's not clear how we get to the final state from various other states (in fact maybe it should say reboot, because that's always how we get to the final state). Going to reboot from a "committed update" will never happen in the case of an error, so this should be reflected somehow. However, from a "non-committed update" it will usually happen.